File: apps/cli/core/pkce.py
import base64
import hashlib
import secrets


def random_b64(nbytes: int = 32) -> str:
    return _b64url(secrets.token_bytes(nbytes))


def pkce_challenge(verifier: str) -> str:
    digest = hashlib.sha256(verifier.encode("ascii")).digest()
    return _b64url(digest)


def _b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")

File: core/src/pkce.ts
import { createHash, randomBytes } from "node:crypto";

export function randomB64(nbytes = 32): string {
  return b64url(randomBytes(nbytes));
}

export function pkceChallenge(verifier: string): string {
  const digest = createHash("sha256").update(verifier, "ascii").digest();
  return b64url(digest);
}

function b64url(data: Buffer): string {
  return data.toString("base64url");
}
